### PR TITLE
[BPK-934] Add BpkIcon support to BpkButton

### PR DESCRIPTION
**Pull request by shaundon**
- Users can now pass in a string to the `icon` prop if they wish. This will be used to make a `BpkIcon` inside the button.
- Users can still pass arbitrary elements in, but they won't be styled or themed, it's up to them.
- We don't validate that the string passed in is a valid icon, because `BpkIcon` does this and the error will bubble up.

Screenshots:

screen shot 2017-10-18 at 17 08 08

screen shot 2017-10-18 at 17 08 41

### CODE REVIEW

File: native/packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap [hunk 1]
```diff
@@ -1257,7 +1351,10 @@ exports[`Android BpkButton should support the "iconOnly" and "large" property 1`
               "paddingLeft": 16,
               "paddingRight": 16,
             },
-            undefined,
+            Object {
```
Comment (matteo-hertel) [inline native/packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap]: what's this bit here? from undefined to an object?

File: native/packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.ios.js.snap [hunk 2]
```diff
@@ -778,7 +809,10 @@ exports[`iOS BpkButton should support having only an icon 1`] = `
               "paddingRight": 12,
               "paddingTop": 8,
             },
-            undefined,
+            Object {
```
Comment (matteo-hertel) [inline native/packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.ios.js.snap]: same as above, what's this bit?

File: native/packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap (hunk 1 above)
Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/__snapshots__/BpkButton-test.android.js.snap]: I'm not 100% but I think it's some styling that wasn't rendered properly before and came out like that, which is working now.

**Comment by k0nserv** [comment]
I prefer passing an Element over the string approach. This is more consistent with web, leaves open the possibility to use arbitrary images(like logos), and removes the need to have `react-native-bpk-component-button` depend on `react-native-bpk-component-icon`. Making it a string is also a breaking change whereas an Element wouldn't be. We still have the option to regress to a string in the future with a breaking change

File: native/packages/react-native-bpk-component-button/src/BpkButton-styles.js [hunk 3]
```diff
@@ -128,11 +128,19 @@ const modifiers = {
...
+      paddingLeft: 0,
+      paddingRight: 0,
+    },
   }),
   iconOnlyLarge: StyleSheet.create({
     container: {
       width: largeHeight,
     },
+    button: {
+      paddingLeft: 0,
```
Comment (k0nserv) [inline native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: `paddingHorizontal` could be used here right?

File: native/packages/react-native-bpk-component-button/src/BpkButton-styles.js [hunk 4]
```diff
@@ -128,11 +128,19 @@ const modifiers = {
     container: {
       width: tokens.spacingXl,
     },
+    button: {
+      paddingLeft: 0,
```
Comment (k0nserv) [inline native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: `paddingHorizontal` could be used here right?

File: native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js [hunk 5]
```diff
@@ -26,8 +26,7 @@ import BpkButton, { propTypes, BUTTON_TYPES } from './BpkButton';
 
 const onPressFn = jest.fn();
 const commonTests = () => {
-  jest.mock('Image', () => 'Image');
-
+  jest.mock('Text', () => 'Text');
```
Comment (k0nserv) [inline native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js]: Why is this line indented like this?

File: native/packages/react-native-bpk-component-button/src/BpkButton-styles.js (hunk 4 above)
Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: Didn't know about that prop, thanks!

File: native/packages/react-native-bpk-component-button/src/BpkButton-styles.js (hunk 3 above)
Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: Just tried it and it didn't have the same effect, despite what the docs said 😢

File: native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js (hunk 5 above)
Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js]: 2 spaces indent because it's in a function - I don't see the issue?

screen shot 2017-10-19 at 11 34 47

File: native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js [hunk 6]
```diff
@@ -16,155 +16,173 @@
...
-          onPress={onPressFn}
-        />,
+       expect(tree).toMatchSnapshot();
+     });
+
+     it('should support elements as icons', () => {
+       const tree = renderer.create(
+         <BpkButton
+           title="Lorem ipsum"
+           icon={<Text>foo</Text>}
```
Comment (matteo-hertel) [inline native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js]: do you think is a good idea to add in the test a behaviour that we don't want to support? I'd use BpkIcon as element here, for people like me that look at the test to see how to use something, it'll encourage to use BpkIcon

File: native/packages/react-native-bpk-component-button/src/BpkButton-styles.js (hunk 3 above)
Comment (k0nserv) [reply native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: Strange

Comment (k0nserv) [reply native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: What happens instead?

Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: Ah. I'm being dumb. Because of the cascade, it was inheriting `paddingLeft` and `paddingRight` from another style. I've changed everywhere to use `paddingHorizontal` and `paddingVertical` and it works like a charm now!

File: native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js (hunk 6 above)
Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js]: I see what you're saying, but then also, we do technically *support* it, we just don't *recommend* it. I'd want to know if a change we made stopped this working correctly, for example.

For me, it's the same as how we have tests for having both text and an icon in the button, even though it's not something we display in the docs.

However if we want to steer users away from this edge case it should be removed from the code sample in `readme.md`, I think.

File: native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js (hunk 5 above)
Comment (k0nserv) [reply native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js]: Hmm sorry, I'm blind it seems

Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js]: It's ok, the glasses you're wearing in your photo do look quite dark 😆

Comment (k0nserv) [reply native/packages/react-native-bpk-component-button/src/BpkButton-test.common.js]: image

File: native/packages/react-native-bpk-component-button/readme.md [hunk 7]
```diff
@@ -47,7 +45,7 @@ export default class App extends Component {
         <BpkButton disabled type="destructive" title={translationHelper.translate('BOOK_FLIGHT')} onPress={() => {}} />
         <BpkButton large type="primary" title={translationHelper.translate('BOOK_FLIGHT')} onPress={() => {}} />
 
-        <BpkButton type="featured" title={translationHelper.translate('BOOK_FLIGHT')} icon={ArrowImage} iconOnly onPress={() => {}} />
+        <BpkButton type="featured" title={translationHelper.translate('BOOK_FLIGHT')} icon="baggage" iconOnly onPress={() => {}} />
```
Comment (k0nserv) [inline native/packages/react-native-bpk-component-button/readme.md]: It would still be good to have a constant for the icons so instead of `"baggage"` we could have `ICONS.BAGGAGE`

File: native/packages/react-native-bpk-component-button/src/BpkButton-styles.js [hunk 8]
```diff
@@ -128,11 +122,17 @@ const modifiers = {
     container: {
       width: tokens.spacingXl,
     },
+    button: {
+      paddingHorizontal: 0,
```
[the remark this answers is not in the document]
Comment (shaundon) [reply native/packages/react-native-bpk-component-button/src/BpkButton-styles.js]: Nah I was wrong - see this comment! https://github.com/Skyscanner/backpack/pull/208#discussion_r145715893

File: native/packages/react-native-bpk-component-button/readme.md (hunk 7 above)
Comment (shaundon) [reply native/packages/react-native-bpk-component-button/readme.md]: Yeah that's a good point. I'm wondering if it should be a separate PR though as it's something `BpkIcon` would export and I don't want to conflate this PR though..

Comment (georgegillams) [reply native/packages/react-native-bpk-component-button/readme.md]: It absolutely should be. Is there a jira card open for this?

Comment (shaundon) [reply native/packages/react-native-bpk-component-button/readme.md]: There is now: https://gojira.skyscanner.net/browse/BPK-1078
